pallets/capacity/src/types.rs

```diff
@@ -0,0 +1,183 @@
+//! Types for the Capacity Pallet
+use super::*;
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::{
+	log::warn, BoundedVec, EqNoBound, PartialEqNoBound, RuntimeDebug, RuntimeDebugNoBound,
+};
+use scale_info::TypeInfo;
+use sp_runtime::traits::{CheckedAdd, CheckedSub, Saturating, Zero};
+
+#[cfg(any(feature = "runtime-benchmarks", test))]
+use sp_std::vec::Vec;
+
+/// The type used for storing information about staking details.
+#[derive(
+	TypeInfo, RuntimeDebugNoBound, PartialEqNoBound, EqNoBound, Clone, Decode, Encode, MaxEncodedLen,
+)]
+#[scale_info(skip_type_params(T))]
+pub struct StakingAccountDetails<T: Config> {
+	/// The amount a Staker has staked, minus the sum of all tokens in `unlocking`.
+	pub active: BalanceOf<T>,
+	/// The total amount of tokens in `active` and `unlocking`
+	pub total: BalanceOf<T>,
+	/// Unstaked balances that are thawing or awaiting withdrawal.
+	pub unlocking: BoundedVec<UnlockChunk<BalanceOf<T>, T::EpochNumber>, T::MaxUnlockingChunks>,
+}
+/// The type that is used to record a single request for a number of tokens to be unlocked.
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug, TypeInfo, MaxEncodedLen)]
+pub struct UnlockChunk<Balance, EpochNumber> {
+	/// Amount to be unlocked.
+	pub value: Balance,
+	/// Block number at which point funds are unlocked.
+	pub thaw_at: EpochNumber,
+}
+
+impl<T: Config> StakingAccountDetails<T> {
+	/// Increases total and active balances by an amount.
+	pub fn increase_by(&mut self, amount: BalanceOf<T>) -> Option<()> {
+		self.total = amount.checked_add(&self.total)?;
+		self.active = amount.checked_add(&self.active)?;
+
+		Some(())
+	}
+
+	/// Calculates a stakable amount from a propose amount.
+	pub fn get_stakable_amount_for(
+		&self,
+		staker: &T::AccountId,
+		proposed_amount: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		let account_balance = T::Currency::free_balance(&staker);
+		let available_staking_balance = account_balance.saturating_sub(self.total);
+		available_staking_balance.min(proposed_amount)
+	}
+
+	#[cfg(any(feature = "runtime-benchmarks", test))]
+	///  tmp fn for testing only
+	/// set unlock chunks with (balance, thaw_at).  does not check that the unlock chunks
+	/// don't exceed total.
+	/// returns true on success, false on failure (?)
+	// TODO: remove this when finished and use production fn
```

Comment:
should this get removed or not yet?

Reply:
I've been meaning to get back to this, thanks for the bump

Reply:
Decided to keep it.